Source/JavaScriptCore/bytecode/ArrayAllocationProfile.h

```diff
@@ -62,17 +62,24 @@ class ArrayAllocationProfile {
     unsigned vectorLengthHint()
     {
         ASSERT(!isCompilationThread());
-        JSArray* lastArray = m_lastArray;
-        if (lastArray && (m_largestSeenVectorLength != BASE_CONTIGUOUS_VECTOR_LEN_MAX) && UNLIKELY(lastArray->getVectorLength() > m_largestSeenVectorLength))
+        IndexingTypeAndVectorLength lastArray = m_lastArray;
+        if (!!lastArray && (m_largestSeenVectorLength != BASE_CONTIGUOUS_VECTOR_LEN_MAX) && UNLIKELY(lastArray.vectorLength() > m_largestSeenVectorLength))
             updateProfile();
         return m_largestSeenVectorLength;
     }
     
     JSArray* updateLastAllocation(JSArray* lastArray)
     {
-        m_lastArray = lastArray;
+        ASSERT(!isCompilationThread());
+        m_lastArray = IndexingTypeAndVectorLength(lastArray->indexingType(), lastArray->getVectorLength());
```

Comment:
The reason why we keep a pointer, and not getting the indexing type & vector length at this point is that, they can be changed later, and we would like to capture the final results.
For example, ArrayAllocation site would see VectorLength = 0 and Undecided shape initially. And while this array is used in the user's program, it will get decided shape (e.g. Int32Shape), and it gets the vector length that is suitable for the user's program. And by putting a pointer here first and reading this data while compiling, we would like to capture this "final shape of the allocated array". So, I think this change is not successfully the final shape of arrays allocated from this allocation site.

Reply:
That's a great point, thanks!